### PR TITLE
Onboarding UI

### PR DESCRIPTION
**Pull request by francisschmaltz**
# Onboarding UI Refinement
Updates to the on boarding UI to bring it inline with mocks in Zeplin

Notes:
- Site admin cards are added when needed.

The onboarding list collapses to a dropdown for status (future use implied) and the user stats are expanded by default.

> Commit d22ab49 adds more interaction to the onboarding. As an admin completes a step, more of the site-admin cards are shown.
For instance: when a user enables a repo they will see the "Explore" and "2 repositories" cards appear. This also solves the problem of hiding the user stats during setup because if an instance has less than 2 users, it won't show the stats.

## User Profile Onboarding:

Screen Shot 2019-03-14 at 9 25 03 PM
Screen Shot 2019-03-14 at 9 25 12 PM

## Onboarding dropdown:

Screen Shot 2019-03-14 at 9 25 29 PM
Screen Shot 2019-03-14 at 9 25 44 PM

## Site Admin Onboarding: 

Screen Shot 2019-03-14 at 9 25 54 PM

## Site Admin Completed Onboarding

Screen Shot 2019-03-14 at 9 37 00 PM
Screen Shot 2019-03-14 at 9 37 07 PM

### CODE REVIEW

**Comment by francisschmaltz** [comment]
Commit 
https://github.com/sourcegraph/sourcegraph/pull/2641/commits/d22ab49e17d3661159526f3b5025e3c2d1cc4b9a adds more interaction to the onboarding. As an admin completes a step, more of the site-admin cards are shown.

For instance: when a user enables a repo they will see the "Explore" and "2 repositories" cards appear. This also solves the problem of hiding the user stats during setup because if an instance has less than 2 users, it won't show the stats.

File: shared/src/components/activation/ActivationChecklist.scss [hunk 1]
```diff
@@ -1,34 +1,84 @@
...
+            bottom: 0;
+        }
 
         @include hover-focus {
             color: $dropdown-link-hover-color;
             text-decoration: none;
-            @include gradient-bg($dropdown-link-hover-bg);
+            background-color: var(--card-cap-bg);
+            margin-bottom: -1px;
+            z-index: 20;
```
Comment (beyang) [inline shared/src/components/activation/ActivationChecklist.scss]:
```suggestion
```

File: shared/src/components/activation/ActivationChecklist.scss [hunk 2]
```diff
@@ -1,34 +1,84 @@
 .activation-checklist {
-    margin-top: 0.5rem;
-
     &__item {
-        border-bottom: 1px solid $dropdown-border-color;
+        box-sizing: content-box;
+        position: relative;
+        margin-top: -1px;
```
Comment (beyang) [inline shared/src/components/activation/ActivationChecklist.scss]: What's the motivation behind having these -1px margins? Could we use something like `box-shadow: inset 0px 0px 0px 1px $color;` instead?

File: shared/src/components/activation/ActivationChecklist.scss [hunk 3]
```diff
@@ -1,34 +1,84 @@
...
-
 .activation-item {
-    &__checkbox {
-        &--done {
-            color: $oc-green-6;
-        }
-        &--todo {
-            color: $dropdown-border-color;
-        }
+    cursor: pointer;
```
Comment (beyang) [inline shared/src/components/activation/ActivationChecklist.scss]: Do we need the `cursor` here? Those items were previously wrapped in a `<LinkOrButton>` item, which took care of adding link-like properties. Why the change?

Comment (francisschmaltz) [reply shared/src/components/activation/ActivationChecklist.scss]: I needed to wrap all items in the same elements so their look is uniform. Previous there was a non-link that was included and it displayed differently. This makes sure that the look is uniform across all items.

File: shared/src/components/activation/ActivationChecklist.scss (hunk 1 above)
Comment (francisschmaltz) [reply shared/src/components/activation/ActivationChecklist.scss]: The z-index is required for the hover UI to prevent the borders from appearing over the element that is selected.

File: shared/src/components/activation/ActivationChecklist.scss (hunk 2 above)
Comment (francisschmaltz) [reply shared/src/components/activation/ActivationChecklist.scss]: The `-1px` margin makes sure that the borders of the list elements overlap.

File: shared/src/components/activation/ActivationDropdown.scss [hunk 4]
```diff
@@ -1,3 +1,91 @@
...
+        z-index: 995;
+        background-position: center;
+        background-size: 110%;
+        background-repeat: no-repeat;
+    }
+
+    &-item {
+        box-sizing: border-box;
+        position: relative;
+        margin-top: -1px;
```
Comment (beyang) [inline shared/src/components/activation/ActivationDropdown.scss]: margin -1px (see above)

File: shared/src/components/activation/ActivationDropdown.scss [hunk 5]
```diff
@@ -1,3 +1,91 @@
...
+        margin-bottom: -1px;
+        padding: 0;
+        z-index: 10;
+        transition: all 250ms ease;
+        white-space: normal;
+        z-index: 996;
```
Comment (beyang) [inline shared/src/components/activation/ActivationDropdown.scss]: why multiple `z-index`?

Comment (francisschmaltz) [reply shared/src/components/activation/ActivationDropdown.scss]: It should be 996, not 10

File: web/src/user/account/UserAccountProfilePage.scss [hunk 6]
```diff
@@ -25,10 +25,32 @@
         margin-top: 1rem;
     }
 
-    &__activation-container {
-        border: 1px solid $dropdown-border-color;
-        border-radius: 3px;
-        margin-bottom: 1rem;
-        padding: 1rem;
+    &__activation {
+        &-container {
```
Comment (beyang) [inline web/src/user/account/UserAccountProfilePage.scss]: I think this violates BEM naming conventions and we'd prefer explicit names for different elements. E.g.,

```
&__activation-container {
  ...
}

&__activation-header {
  ...
}
```

File: shared/src/components/activation/ActivationChecklist.scss (hunk 1 above)
Comment (beyang) [reply shared/src/components/activation/ActivationChecklist.scss]: Why are there 2 `z-index` fields?

**Comment by beyang** [comment]
+1 from me, but @felixfbecker can you take a look, too?

File: shared/src/components/activation/ActivationDropdown.scss (hunk 4 above)
Comment (francisschmaltz) [reply shared/src/components/activation/ActivationDropdown.scss]: The `-1px` margin makes sure that the borders of the list elements overlap.

File: shared/src/components/activation/ActivationChecklist.scss (hunk 1 above)
Comment (francisschmaltz) [reply shared/src/components/activation/ActivationChecklist.scss]: The z-index increases when the item is hovered so it appears above the borders of elements on top.

**Comment by francisschmaltz** [comment]
Before/After of popover updates:

Before:

Screen Shot 2019-03-15 at 3 34 27 PM

After:

Screen Shot 2019-03-15 at 3 34 22 PM

File: web/src/components/PopoverButton.scss [hunk 7]
```diff
@@ -53,12 +56,15 @@
         // Flip vertically so that it's a caret-up icon (not caret-down).
         transform: rotate(180deg) translateY(-1px);
     }
+    &--open#{&}__btn {
```
Comment (felixfbecker) [inline web/src/components/PopoverButton.scss]: This looks weird, what does this expand too?

File: web/src/user/account/UserAccountProfilePage.scss [hunk 8]
```diff
@@ -29,6 +29,26 @@
...
+        overflow: hidden;
+    }
+
+    &__activation-header {
+        padding: 0.5rem 1rem;
+        white-space: normal;
+    }
+    &__activation-title {
+        position: relative;
+        font-size: 18px;
```
Comment (felixfbecker) [inline web/src/user/account/UserAccountProfilePage.scss]: I see a lot of explicit font size overrides in this PR which strikes me as an anti-pattern for consistency

File: shared/src/components/activation/ActivationChecklist.scss [hunk 9]
```diff
@@ -1,34 +1,84 @@
...
+    // Padding-right prevents text from overlapping icon.
+    padding: 0.5rem 2.25rem 0.5rem 1.25rem;
+
+    &__checkbox--done {
+        color: $oc-green-6;
+        position: absolute;
+        right: 1rem;
+        top: calc(50% - 10.5px);
```
[the remark this answers is not in the document]
Comment (francisschmaltz) [reply shared/src/components/activation/ActivationChecklist.scss]: It does some strange tricks on @2x and @3x screens to make it perfectly centered.

File: web/src/components/PopoverButton.scss (hunk 7 above)
Comment (francisschmaltz) [reply web/src/components/PopoverButton.scss]: This changes the z-index only on `btn` when the popover is open. So the shadow doesn't appear over the popover button.

File: shared/src/components/activation/ActivationChecklist.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply shared/src/components/activation/ActivationChecklist.scss]: Have you considered using flexbox?

File: shared/src/components/activation/ActivationDropdown.scss [hunk 10]
```diff
@@ -1,3 +1,91 @@
...
+
+    &-background {
+        position: absolute;
+        top: 0;
+        left: 0;
+        right: 0;
+        bottom: 0;
+        background-image: -webkit-linear-gradient(top, rgba(14, 18, 27, 0), rgba(14, 18, 27, 1)),
+            url(/.assets/img/onboarding-bg-dark.png);
+        background-image: -moz-linear-gradient(top, rgba(14, 18, 27, 0), rgba(14, 18, 27, 1)),
```
[the remark this answers is not in the document]
Comment (francisschmaltz) [reply shared/src/components/activation/ActivationDropdown.scss]: Autoprefixer did not add these (I'm guessing because it's on background-image and I use both an image and a gradient.

File: web/src/components/PopoverButton.scss (hunk 7 above)
Comment (felixfbecker) [reply web/src/components/PopoverButton.scss]: So what does this selector expand to? I can't get it to expand to anything in Sass playground:

image

File: shared/src/components/activation/ActivationDropdown.scss (hunk 10 above)
[the remark this answers is not in the document]
Comment (francisschmaltz) [reply shared/src/components/activation/ActivationDropdown.scss]: https://github.com/sourcegraph/sourcegraph/issues/2776

File: web/src/components/PopoverButton.scss (hunk 7 above)
Comment (francisschmaltz) [reply web/src/components/PopoverButton.scss]: It should compile to `.popover-button--open.popover-button__btn`

Comment (francisschmaltz) [reply web/src/components/PopoverButton.scss]: `#{&}` should make the text value of the parent element.

File: shared/src/components/activation/ActivationDropdown.scss (hunk 10 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply shared/src/components/activation/ActivationDropdown.scss]: Autoprefixer is capable of adding them:

image

If they are not added, that means they are supported in all the browsers we target and they should be removed, or our browserlist config should be corrected (in the future, please follow these steps).

image

https://caniuse.com/#feat=css-gradients

File: web/src/components/PopoverButton.scss (hunk 7 above)
Comment (francisschmaltz) [reply web/src/components/PopoverButton.scss]: Screen Shot 2019-03-15 at 4 57 48 PM
